lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +54,142 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    ch = {
+      contextPreviewExists: false
+
+      removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while editor.session.getContext(i - 1) == "equation"
+          i -= 1
+        start = i
+        while editor.session.getContext(i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(ch.removeRegex, "")
+
+      getTopmostRowNumber: ->
+        parseInt(jqEditorContainer.find("div.ace_gutter > div.ace_layer.ace_gutter-layer.ace_folding-enabled > div:nth-child(1)").text())
+
+      getPopoverPosition: (row) ->
+        secondRowSelector = "div.ace_scroller > div > div.ace_layer.ace_text-layer > div:nth-child(2)"
```

Comment:
You may want to try [`editor.renderer.textToScreenCoordinates`](https://github.com/bardsoftware/ace/blob/internship-2016/lib/ace/virtual_renderer.js#L1434)